File: 7/js/euler7.js
const isPrime = (x) => {
	if (x < 2) {
		return false;
	}
	let divisor = 2|0;
	while (divisor * divisor <= x) {
		if (!(x % divisor)) {
			return false;
		}
		divisor += 1;
	}

	return true;
};

const countPrimes = (n) => {
	let x = 2|0;
	for(;n; x += 1) {
		if (isPrime(x)) {
			n -= 1;
		}
	}

	return x-1;
};

const getPrime = (n) => countPrimes(n|0);

const myArgs = process.argv.slice(2);
const [n] = myArgs;
const startTime = Date.now() / 1000;
prime = getPrime(n);
const endTime = Date.now() / 1000;
console.log('Prime: ', prime);
console.log('Time:', endTime - startTime);

File: 7/erlang/euler7.erl
-module(euler7).

-export([main/1]).

main([Arg| _]) ->
    {N, _} = string:to_integer(Arg),
    StartTime = erlang:timestamp(),
    Prime = get_prime(N),
    EndTime = erlang:timestamp(),
    io:format("~B prime : ~B~n", [N, Prime]),
    io:format("Time : ~f~n", [timer:now_diff(EndTime, StartTime)/1000000]),
    erlang:halt().

get_prime(N) -> count_primes(2, N).

count_primes(X, 0) -> X - 1;
count_primes(X, N) -> count_primes(X + 1, decIf(N, is_prime(X, 2))).

decIf(N, P) when P  -> N - 1;
decIf(N, _)         -> N.

is_prime( X, _Div) when X < 2           -> false;
is_prime(_X,  Div) when Div < 2         -> true;
is_prime(X, Div)                        -> is_prime_optimized(X, Div).


is_prime_optimized( X,  Div) when Div * Div > X   -> true;
is_prime_optimized( X,  Div) when X rem Div == 0  -> false;
is_prime_optimized( X,  Div)                      -> is_prime_optimized(X, Div + 1).
